include/mbgl/style/light.hpp

```diff
@@ -0,0 +1,294 @@
+#pragma once
+
+#include <mbgl/style/types.hpp>
+#include <mbgl/style/property_value.hpp>
+#include <mbgl/style/property_evaluator.hpp>
+#include <mbgl/style/property_evaluation_parameters.hpp>
+#include <mbgl/style/transition_options.hpp>
+#include <mbgl/style/cascade_parameters.hpp>
+#include <mbgl/util/constants.hpp>
+#include <mbgl/util/color.hpp>
+#include <mbgl/util/constants.hpp>
+#include <mbgl/util/interpolate.hpp>
+#include <mbgl/util/indexed_tuple.hpp>
+#include <mbgl/util/ignore.hpp>
+
+#include <memory>
+
+namespace mbgl {
+namespace style {
+
+template <class Value>
+class UnevaluatedLightProperty {
+public:
+    UnevaluatedLightProperty() = default;
+
+    UnevaluatedLightProperty(Value value_,
+                             UnevaluatedLightProperty<Value> prior_,
+                             TransitionOptions transition,
+                             TimePoint now)
+        : begin(now + transition.delay.value_or(Duration::zero())),
+          end(begin + transition.duration.value_or(Duration::zero())),
+          value(std::move(value_)) {
+        if (transition) {
+            prior = { std::move(prior_) };
+        }
+    }
+
+    template <class Evaluator>
+    auto evaluate(const Evaluator& evaluator, TimePoint now) {
+        auto finalValue = value.evaluate(evaluator);
+        if (!prior) {
+            // No prior value.
+            return finalValue;
+        } else if (now >= end) {
+            // Transition from prior value is now complete.
+            prior = {};
+            return finalValue;
+        } else if (now < begin) {
+            // Transition hasn't started yet.
+            return prior->get().evaluate(evaluator, now);
+        } else {
+            // Interpolate between recursively-calculated prior value and final.
+            float t = std::chrono::duration<float>(now - begin) / (end - begin);
+            return util::interpolate(prior->get().evaluate(evaluator, now), finalValue,
+                                     util::DEFAULT_TRANSITION_EASE.solve(t, 0.001));
+        }
+    }
+
+    bool hasTransition() const {
+        return bool(prior);
+    }
+
+    bool isUndefined() const {
+        return value.isUndefined();
+    }
+
+    const Value& getValue() const {
+        return value;
+    }
+
+private:
+    optional<mapbox::util::recursive_wrapper<UnevaluatedLightProperty<Value>>> prior;
+    TimePoint begin;
+    TimePoint end;
+    Value value;
+};
+
+template <class Value>
+class CascadingLightProperty {
```

Comment:
Per chat, light properties support transitions but not style classes, so this class can be simplified, and "Cascading" removed from the name. It just serves to tie the `Value` and `TransitionOptions` together.